File: apps/tilawa/lib/features/islamic_widgets/domain/entities/ayah_widget_payload.dart
/// Display-ready content for the Ayah of the Day widget (spec 041, US2).
///
/// The verse itself ships as pre-rendered PNG artifacts (one per theme) so the
/// native provider shows authentic QCF glyphs without a Dart isolate. Paths
/// point inside the app's own files directory; no user content or location
/// data crosses the bridge (privacy contract of the widget envelope).
class AyahWidgetPayload {
  const AyahWidgetPayload({
    required this.dateKey,
    required this.surahNumber,
    required this.ayahNumber,
    required this.pageNumber,
    required this.caption,
    required this.imagePathLight,
    required this.imagePathDark,
  });

  /// Local calendar day this selection belongs to (yyyy-MM-dd).
  final String dateKey;

  final int surahNumber;
  final int ayahNumber;

  /// Madinah Mushaf page opened when the widget is tapped.
  final int pageNumber;

  /// Localized reference line, e.g. "سورة البقرة · ١٥٢".
  final String caption;

  /// Absolute paths of the rendered QCF verse artifacts.
  final String imagePathLight;
  final String imagePathDark;

  Map<String, Object?> toJson() => <String, Object?>{
    'dateKey': dateKey,
    'surahNumber': surahNumber,
    'ayahNumber': ayahNumber,
    'pageNumber': pageNumber,
    'caption': caption,
    'imagePathLight': imagePathLight,
    'imagePathDark': imagePathDark,
  };
}

File: apps/tilawa/android/app/src/main/kotlin/com/tilawa/app/widget/ayah/AyahWidgetPayload.kt
package com.tilawa.app.widget.ayah

import org.json.JSONObject

/**
 * Parsed payload of an Ayah of the Day snapshot (spec 041, US2). The verse
 * arrives as pre-rendered QCF PNG artifacts (one per theme) written by the
 * Flutter repository into the app's files directory.
 */
internal data class AyahWidgetPayload(
    val dateKey: String,
    val surahNumber: Int,
    val ayahNumber: Int,
    val pageNumber: Int,
    val caption: String,
    val imagePathLight: String,
    val imagePathDark: String,
) {
    companion object {
        /** Returns null when any required field is missing or invalid. */
        fun parse(payload: JSONObject): AyahWidgetPayload? {
            val dateKey = payload.optString("dateKey")
            val surah = payload.optInt("surahNumber", 0)
            val ayah = payload.optInt("ayahNumber", 0)
            val page = payload.optInt("pageNumber", 0)
            val caption = payload.optString("caption")
            val light = payload.optString("imagePathLight")
            val dark = payload.optString("imagePathDark")
            if (dateKey.isBlank() || light.isBlank() || dark.isBlank()) return null
            if (surah !in 1..114 || ayah < 1 || page !in 1..604) return null
            return AyahWidgetPayload(
                dateKey = dateKey,
                surahNumber = surah,
                ayahNumber = ayah,
                pageNumber = page,
                caption = caption,
                imagePathLight = light,
                imagePathDark = dark,
            )
        }
    }
}
